### PR TITLE
Tip 1073: use new product value factories

### PR DESCRIPTION
**Pull request by ahocquard**
**Description (for Contributor and Core Developer)**

Goal of this PR is to only use new factories for the values. These factories help for the performance. A big amount of work was done previously to do it (do not accept null being the major one).

What I did in this PR:

- [x] pre-validation of the scopability/localizability when we create a value according to the attribute type (PHP error otherwise)

=> ideally, it should be done in validators but it was a too big step

- [x] filter values about deactivated locale/ removed channel 

- [x] I take profit of this PR to have a single service for filtering non existent data. It's not splitted anymore in 4 services called in the value factory. It means we can use it easily in some other parts of the PIM using raw values but not value collection. 
Better in term of responsibility also.

- [x] I removed a lot of useless behat testing the same things 3/4 times or skipped since 4 years

- [x] I removed dead code that I had in mind

What's left:

- [ ] Remove previous value factory (in another PR)
- [ ] Prevent null values in some new value factories
- [ ] Move some factory classes in the right directory
- [ ] Create ValueCollection without filtering data for the indexation
- [ ] Do not filter empty values for completeness calculation as we don't store anymore null values
- [ ] Fix dirty workaround with lazy services with the Symfony 4 container + oauth 

Et voilà!

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/EmptyValuesCleaner.php [hunk 1]
```diff
@@ -4,14 +4,28 @@
...
  * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
  */
 final class EmptyValuesCleaner
 {
+    /**
+     * @param array $rawValueCollections array of raw value collections
+     * It allows to filter several raw values at once from different product or product models.
+     * [
+     *    ['attribute_code' => [...], 'attribute_code_2' => [...]],
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/EmptyValuesCleaner.php]: wrong comment, it's indexed by product identifier

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/EmptyValuesCleaner.php [hunk 2]
```diff
@@ -4,14 +4,28 @@
 namespace Akeneo\Pim\Enrichment\Component\Product\Factory;
 
 /**
- * This service removes all empty values from a rawValueCollections
+ * This service removes all empty values from raw values stored in the database.
+ * Do note that we don't store null values or empty arrays in the database. However, the consistency of the data in not guaranteed
+ * in the JSON of the raw values.
+ *
+ * Therefore, when filtering on data non existing data, we can have values with empty arrays or empty data.
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/EmptyValuesCleaner.php]:
```suggestion
 * Therefore, when filtering on non existing data, we can have values with empty arrays or empty data.
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/NonExistentValuesFilter/ChainedNonExistentValuesFilter.php [hunk 3]
```diff
@@ -4,23 +4,58 @@
 
 namespace Akeneo\Pim\Enrichment\Component\Product\Factory\NonExistentValuesFilter;
 
+use Akeneo\Pim\Enrichment\Component\Product\Factory\EmptyValuesCleaner;
+use Akeneo\Pim\Enrichment\Component\Product\Factory\TransformRawValuesCollections;
+use Akeneo\Tool\Component\StorageUtils\Repository\IdentifiableObjectRepositoryInterface;
+
 /**
+ * The implementation of this non existent filter use a pivot format internally.
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/NonExistentValuesFilter/ChainedNonExistentValuesFilter.php]:
```suggestion
 * The implementation of this non existent filter uses a pivot format internally.
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/Value/ValidateAttribute.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * It does not validate the existence or the consistency of the locale according to the channel, as this check
+ * should be done in validation layer.
+ *
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class ValidateAttribute
+{
+    public static function validate(Attribute $attribute, ?string $channelCode, ?string $localeCode): void
```
[the remark this answers is not in the document]
Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Component/Product/Factory/Value/ValidateAttribute.php]: I disagree on that point: if it does not use a service, a static class is perfectly fine. It's like an assert. And it forces to not use a double for that.

File: src/Akeneo/Pim/Enrichment/Component/Product/Builder/EntityWithValuesBuilder.php [hunk 5]
```diff
@@ -23,28 +25,21 @@ class EntityWithValuesBuilder implements EntityWithValuesBuilderInterface
...
 
+    /** @var GetAttributes */
+    protected $getAttributesQuery;
+
     /**
      * @param AttributeValuesResolverInterface $valuesResolver
      * @param ValueFactory                     $productValueFactory
      */
     public function __construct(
         AttributeValuesResolverInterface $valuesResolver,
```
[the remark this answers is not in the document]
Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Component/Product/Builder/EntityWithValuesBuilder.php]: :+1: I will remove it in next PR

**Pull request merged**